File: variant-a/java-service/src/main/java/com/flashsale/api/service/AdaptiveInventoryManager.java
package com.flashsale.api.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ClassPathResource;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.script.DefaultRedisScript;
import org.springframework.scripting.support.ResourceScriptSource;
import org.springframework.stereotype.Service;

import jakarta.annotation.PostConstruct;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Manages adaptive inventory services for multiple SKUs
 * Singleton pattern - one instance per application
 */
@Service
public class AdaptiveInventoryManager {

    private static final Logger logger = LoggerFactory.getLogger(AdaptiveInventoryManager.class);

    private final RedisTemplate<String, Object> redisTemplate;
    private DefaultRedisScript<Long> luaScript;
    private final Map<String, AdaptiveInventoryService> services = new ConcurrentHashMap<>();

    @Autowired
    public AdaptiveInventoryManager(RedisTemplate<String, Object> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    /**
     * Initialize and load Lua script into Redis
     */
    @PostConstruct
    public void initialize() {
        try {
            luaScript = new DefaultRedisScript<>();
            luaScript.setResultType(Long.class);

            // Load Lua script from classpath
            ClassPathResource scriptResource = new ClassPathResource("lua/inventory_refill.lua");
            ResourceScriptSource scriptSource = new ResourceScriptSource(scriptResource);
            luaScript.setScriptSource(scriptSource);

            // Load script content to get SHA (for logging)
            String scriptContent = scriptSource.getScriptAsString();
            String sha = redisTemplate.execute((org.springframework.data.redis.core.RedisCallback<String>) connection ->
                new String(connection.scriptingCommands().scriptLoad(scriptContent.getBytes()))
            );

            logger.info("AdaptiveInventoryManager initialized");
            logger.info("  Lua Script SHA: {}", sha);
        } catch (IOException e) {
            logger.error("Failed to load Lua script", e);
            throw new RuntimeException("Failed to initialize AdaptiveInventoryManager", e);
        }
    }

    /**
     * Get or create adaptive inventory service for a SKU
     *
     * @param campaignId Campaign ID
     * @param skuId SKU ID
     * @return AdaptiveInventoryService instance for this SKU
     */
    public AdaptiveInventoryService getService(String campaignId, String skuId) {
        if (luaScript == null) {
            throw new IllegalStateException("AdaptiveInventoryManager not initialized");
        }

        String serviceKey = campaignId + ":" + skuId;

        return services.computeIfAbsent(serviceKey, key -> {
            logger.debug("Creating new AdaptiveInventoryService for {}", serviceKey);
            return new AdaptiveInventoryService(
                campaignId,
                skuId,
                redisTemplate,
                luaScript
            );
        });
    }

    /**
     * Get metrics for all active services
     */
    public Map<String, Map<String, Object>> getAllMetrics() {
        Map<String, Map<String, Object>> metrics = new HashMap<>();

        for (Map.Entry<String, AdaptiveInventoryService> entry : services.entrySet()) {
            AdaptiveInventoryService service = entry.getValue();
            AdaptiveInventoryService.InventoryMetrics m = service.getMetrics();

            Map<String, Object> serviceMetrics = new HashMap<>();
            serviceMetrics.put("batch_mode_requests", m.batchModeRequests);
            serviceMetrics.put("direct_mode_requests", m.directModeRequests);
            serviceMetrics.put("total_redis_calls", m.totalRedisCalls);
            serviceMetrics.put("mode_switched", m.modeSwitched);
            serviceMetrics.put("current_mode", service.getCurrentMode());

            metrics.put(entry.getKey(), serviceMetrics);
        }

        return metrics;
    }

    /**
     * Get count of active services
     */
    public int getServiceCount() {
        return services.size();
    }
}

File: variant-a/csharp-service/Services/AdaptiveInventoryManager.cs
using StackExchange.Redis;
using System.Collections.Concurrent;

namespace FlashSale.Api.Services;

/// <summary>
/// Manages adaptive inventory services for multiple SKUs
/// Singleton pattern - one instance per application
/// </summary>
public class AdaptiveInventoryManager
{
    private readonly ILogger<AdaptiveInventoryManager> _logger;
    private readonly IDatabase _db;
    private string? _luaScriptSha;
    private readonly ConcurrentDictionary<string, AdaptiveInventoryService> _services = new();
    private readonly SemaphoreSlim _initLock = new(1, 1);

    public AdaptiveInventoryManager(ILogger<AdaptiveInventoryManager> logger, IConnectionMultiplexer redis)
    {
        _logger = logger;
        _db = redis.GetDatabase();
    }

    /// <summary>
    /// Initialize and load Lua script into Redis
    /// </summary>
    public async Task InitializeAsync()
    {
        if (_luaScriptSha != null) return;

        await _initLock.WaitAsync();
        try
        {
            if (_luaScriptSha != null) return;

            // Load Lua script from file
            string scriptPath = Path.Combine(AppContext.BaseDirectory, "lua", "inventory_refill.lua");
            string scriptContent = await File.ReadAllTextAsync(scriptPath);

            // Load script into Redis
            var server = _db.Multiplexer.GetServer(_db.Multiplexer.GetEndPoints().First());
            var sha = await server.ScriptLoadAsync(scriptContent);
            _luaScriptSha = BitConverter.ToString(sha).Replace("-", "").ToLower();

            _logger.LogInformation("AdaptiveInventoryManager initialized");
            _logger.LogInformation("  Lua Script SHA: {LuaScriptSha}", _luaScriptSha);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to initialize AdaptiveInventoryManager");
            throw;
        }
        finally
        {
            _initLock.Release();
        }
    }

    /// <summary>
    /// Get or create adaptive inventory service for a SKU
    /// </summary>
    /// <param name="campaignId">Campaign ID</param>
    /// <param name="skuId">SKU ID</param>
    /// <returns>AdaptiveInventoryService instance for this SKU</returns>
    public async Task<AdaptiveInventoryService> GetServiceAsync(string campaignId, string skuId)
    {
        if (_luaScriptSha == null)
        {
            await InitializeAsync();
        }

        string serviceKey = $"{campaignId}:{skuId}";

        return _services.GetOrAdd(serviceKey, key =>
        {
            _logger.LogDebug("Creating new AdaptiveInventoryService for {ServiceKey}", serviceKey);
            return new AdaptiveInventoryService(
                campaignId,
                skuId,
                _db,
                _luaScriptSha!
            );
        });
    }

    /// <summary>
    /// Get metrics for all active services
    /// </summary>
    public Dictionary<string, Dictionary<string, object>> GetAllMetrics()
    {
        var metrics = new Dictionary<string, Dictionary<string, object>>();

        foreach (var entry in _services)
        {
            var m = entry.Value.GetMetrics();

            metrics[entry.Key] = new Dictionary<string, object>
            {
                ["batch_mode_requests"] = m.BatchModeRequests,
                ["direct_mode_requests"] = m.DirectModeRequests,
                ["total_redis_calls"] = m.TotalRedisCalls,
                ["mode_switched"] = m.ModeSwitched,
                ["current_mode"] = entry.Value.CurrentMode
            };
        }

        return metrics;
    }

    /// <summary>
    /// Get count of active services
    /// </summary>
    public int GetServiceCount() => _services.Count;
}
